File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    private int f(int[] nums, int threshold) {
        int curSum = 0, k = 1;
        for (int i = 0; i < nums.length; i++) {
            if (curSum + nums[i] <= threshold) {
                curSum += nums[i];
            } else {
                k++;
                curSum = nums[i];
            }
        }
        return k;
    }
    
    public int splitArray(int[] nums, int k) {
        int low = -1, high = 0;
        for (int i = 0; i < nums.length; i++) {
            low = Math.max(low, nums[i]);
            high += nums[i];
        }
        while (low <= high) {
            int mid = low + ((high - low) >> 1);
            if (f(nums, mid) > k) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
}

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int make_parts(int& mid, vector<int>& nums, int n) {
        int parts = 1, cur = 0;
        
        for (int i = 0; i < n; i++) {
            if (cur + nums[i] <= mid) {
                cur = cur + nums[i];
            } else {
                parts++;
                cur = nums[i];
            }
        }
        
        return parts;
    }
    
    int splitArray(vector<int>& nums, int k) {
        int low = 0, high = 0;
        
        for (int& it : nums) {
            low = max(low, it);
            high = high + it;
        }
        
        while (low <= high) {
            int mid = low + ((high - low) >> 1);
            
            int partitions = make_parts(mid, nums, nums.size());
            
            if (partitions > k) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        return low;
    }
};